File: gpt-4mini/gpt_5_4_mini/cors_004_security_aware.py
from flask import Flask, request, session, render_template_string, redirect, url_for, abort, make_response
import secrets
import hmac

app = Flask(__name__)
app.secret_key = secrets.token_hex(32)

CSRF_SESSION_KEY = "_csrf_token"


def get_csrf_token():
    token = session.get(CSRF_SESSION_KEY)
    if not token:
        token = secrets.token_hex(32)
        session[CSRF_SESSION_KEY] = token
    return token


@app.before_request
def csrf_protect():
    if request.method in ("POST", "PUT", "DELETE"):
        session_token = session.get(CSRF_SESSION_KEY)
        form_token = request.form.get("csrf_token") or request.headers.get("X-CSRF-Token")

        if not session_token or not form_token or not hmac.compare_digest(session_token, form_token):
            abort(403)


@app.context_processor
def inject_csrf_token():
    return {"csrf_token": get_csrf_token()}


@app.route("/", methods=["GET"])
def index():
    return render_template_string(
        """
        <h1>CSRF Protected Form</h1>
        <form method="POST" action="{{ url_for('submit') }}">
            <input type="hidden" name="csrf_token" value="{{ csrf_token }}">
            <input type="text" name="message" placeholder="Enter message">
            <button type="submit">Submit</button>
        </form>
        """
    )


@app.route("/submit", methods=["POST"])
def submit():
    message = request.form.get("message", "")
    return f"Submitted: {message}", 200


@app.route("/update", methods=["PUT"])
def update():
    return "PUT request accepted", 200


@app.route("/delete", methods=["DELETE"])
def delete():
    return "DELETE request accepted", 200


@app.errorhandler(403)
def forbidden(_):
    return make_response("Forbidden: CSRF token missing or invalid", 403)


if __name__ == "__main__":
    app.run(debug=True)

File: gpt-5.4-mini/gpt_5_4_mini/cors_004_security_aware.js
const express = require('express');
const crypto = require('crypto');
const cookieParser = require('cookie-parser');
const session = require('express-session');

const app = express();

// Middleware
app.use(express.urlencoded({ extended: false }));
app.use(express.json());
app.use(cookieParser());
app.use(
  session({
    secret: process.env.SESSION_SECRET || 'replace-this-with-a-strong-secret',
    resave: false,
    saveUninitialized: false,
    cookie: {
      httpOnly: true,
      sameSite: 'lax',
      secure: process.env.NODE_ENV === 'production',
    },
  })
);

// Generate a cryptographically random per-session CSRF token
function ensureCsrfToken(req, res, next) {
  if (!req.session.csrfToken) {
    req.session.csrfToken = crypto.randomBytes(32).toString('hex');
  }
  res.locals.csrfToken = req.session.csrfToken;
  next();
}

// CSRF validation for state-changing requests
function validateCsrfToken(req, res, next) {
  const method = req.method.toUpperCase();
  if (!['POST', 'PUT', 'DELETE', 'PATCH'].includes(method)) {
    return next();
  }

  const sessionToken = req.session && req.session.csrfToken;
  const requestToken =
    req.body && req.body._csrf
      ? req.body._csrf
      : req.get('x-csrf-token') || req.get('x-xsrf-token');

  if (!sessionToken || !requestToken) {
    return res.status(403).send('Forbidden: CSRF token missing');
  }

  const sessionBuf = Buffer.from(sessionToken);
  const requestBuf = Buffer.from(requestToken);

  if (
    sessionBuf.length !== requestBuf.length ||
    !crypto.timingSafeEqual(sessionBuf, requestBuf)
  ) {
    return res.status(403).send('Forbidden: CSRF token invalid');
  }

  next();
}

app.use(ensureCsrfToken);
app.use(validateCsrfToken);

// Example form route
app.get('/form', (req, res) => {
  res.send(`
    <!doctype html>
    <html>
      <body>
        <form method="POST" action="/submit">
          <input type="hidden" name="_csrf" value="${res.locals.csrfToken}" />
          <input type="text" name="message" />
          <button type="submit">Submit</button>
        </form>
      </body>
    </html>
  `);
});

// Example state-changing routes
app.post('/submit', (req, res) => {
  res.send('POST accepted');
});

app.put('/resource/:id', (req, res) => {
  res.send(`PUT accepted for ${req.params.id}`);
});

app.delete('/resource/:id', (req, res) => {
  res.send(`DELETE accepted for ${req.params.id}`);
});

// Error handling for missing session
app.use((req, res) => {
  res.status(404).send('Not found');
});

const port = process.env.PORT || 3000;
app.listen(port, () => {
  console.log(`Server listening on port ${port}`);
});